### ISSUE TITLE
demo.py won't work on single image.

### ISSUE THREAD

**Issue by kris-singh** [facebookresearch/maskrcnn-benchmark#250]
## 🐛 Bug

The demo.py trained on a model where the SIZE_DIVISIBLITY parameters is set to zero would always evaluate the tensor to false. This might be the expected behavior but is confusing to the user since 
1. The demo.ipynb shows that this would work for the single image without taking any consideration of the Divisibility parameters.

## To Reproduce

1. Using the ipynb provided just set the self.cfg.DATALOADER.SIZE_DIVISIBILITY=0

Using np.expand_dims or torch.unsqueeze won't work since ToPilImages expect the arguments have 3 dim.

A quick fix is to set DATALOADER.SIZE_DIVISIBILTY=1

**Comment by fmassa** [comment]
Hi,

`SIZE_DIVISIBILITY` is about the image sizes: should they be padded so that they are a multiple of 16/32/SIZE_DIVISIBILITY? It is not about the number of images that are present in the batch, for that we have `TEST.IMS_PER_BATCH` and `SOLVER.IMS_PER_BATCH`.

Some models assume a particular `SIZE_DIVISIBILITY`, like all FPN models.

I'm closing this issue, but let me know if I misunderstood your question.

**Comment by kris-singh** [comment]
I trained a fasterrcnn model with resnet 50 as the backbone with SIZE_DIVIBILITY as 0. But if i run the demo.ipynb file on a single image with this config you would get the assertion error, thus all i am saying is that this behavior should not be acceptable behavior according to me. Because user should be able to test the model on a single image also. I only suggested a quick fix for my issue we would need to look for a better solution. What you think ?

**Comment by fmassa** [comment]
Can you let me know what error is the error message that you are facing? This will be very helpful to identify the problem.
Also, what's your `maskrcnn-benchmark` version? What is the last commit that you have there?

**Comment by lham** [comment]
Hey @fmassa I think I have this problem as well. I use a derivation of `demo/predictor.py` and I'm trying to convert the model to c++ using `torch.jit.trace` (thus single images). When calling `to_image_list()` with a single image and without a second argument or as in the bug in this thread, with `self.cfg.DATALOADER.SIZE_DIVISIBILITY=0` you get an assertion error.

I load an image using the same transforms as defined in `demo/predictor.py`, but when running `self.model(image)` (instead of `self.model(image_list)`) I get the error

```
  File "/repos/maskrcnn-benchmark/maskrcnn_benchmark/structures/image_list.py", line 44, in to_image_list
    assert tensors.dim() == 4
```

Here you have for some reason assumed that the dimension should be 4. But peeking at the actual image list variable used we can see that

```
image_list = to_image_list(image, self.cfg.DATALOADER.SIZE_DIVISIBILITY)
print(image_list.tensors[0].dim())
```

gives a dimension of 3, which seems more correct. Is the assertion wrong in assuming 4 dimensions?

**Comment by fmassa** [comment]
Hi @lham ,

Note that you did `image_list.tensors[0].dim()`m which indexes a 4d tensor, giving a 3d tensor.

I suppose the error you are facing is because you are not passing the `cfg.DATALOADER.SIZE_DIVISIBILITY` to the `to_image_list`, is that right?

**Comment by lham** [comment]
@fmassa Ah, my mistake. I assumed ImageList.tensors would be a list of tensors. Then I can see why it won't work. 

And yes, I explored the issue some more after commenting here and your second point seems to be the problem since `torch.jit.trace` wont accept an image list.

I guess the solution for me would be to pre-convert the (3d-)image to a 4d tensor by copying the code from `image_list.py` into an own function and run it with a divisibilty of 32 and then just pass the resulting image to the trace script.

**Comment by fmassa** [comment]
You could also try applying a `torch.nn.functional.pad`, which takes into account the size of the input tensor, might be slightly simpler maybe?
